File: WassersteinBayesRepulsiveGMM/src/numerical_Zhat.jl
function numerical_Zhat(Mu_mc, Sig_mc, g₀, ℓ, t_max)::Vector{Float64}
    n_mc = size(Mu_mc, 3) 
    gg = zeros(Float64, t_max, n_mc)

    # println(size(Mu_mc), " ", size(Sig_mc), "....", ℓ+t_max-1)
    @inbounds for n = 1:n_mc, k = ℓ:(ℓ+t_max-1)  
        min_d = 1.
        @inbounds for i = 1:(k-1), j = (i+1):k
            # println(Sig_mc[:, :, i, n],  Sig_mc[:, :, j, n])
            d = wass_gauss(
                Mu_mc[:, i, n], Sig_mc[:, :, i, n], 
                Mu_mc[:, j, n], Sig_mc[:, :, j, n]) 
            min_d = min(min_d, d/(g₀+d)) 
        end
        gg[k-ℓ+1, n] = min_d  
    end 
    Ẑ = log.(mean(gg; dims=2)) |> vec

    return Ẑ
end 

File: pywbrgmm/numerical_Zhat.py
import jax.numpy as jnp 


def numerical_Zhat(Mu_mc, Sig_mc, g0, ell, t_max): 
    n_mc = Mu_mc.shape[0] 
    gg = jnp.zeros(n_mc, t_max)
 
    for n in range(n_mc): 
        for k in range(ell, (ell+t_max)):  
            min_d = 1.
            for i in range(k):
                for j in range(i:k-1):
                # println(Sig_mc[:, :, i, n],  Sig_mc[:, :, j, n])
                d = wass_gauss(
                    Mu_mc[:, i, n], Sig_mc[:, :, i, n], 
                    Mu_mc[:, j, n], Sig_mc[:, :, j, n]) 
                min_d = min(min_d, d/(g₀+d)) 
            gg[n, k-ell] = min_d  

    Zhat = jnp.mean(gg, axis=1).log()
    return Zhat 